### PR TITLE
Es private methods and accessors simplify emit

### PR DESCRIPTION
**Pull request by dragomirtitian**

### CODE REVIEW

File: src/compiler/factory/emitHelpers.ts [hunk 1]
```diff
@@ -32,21 +32,8 @@ namespace ts {
...
-        // Class Static Private Helpers
-        createClassStaticPrivateFieldGetHelper(receiver: Expression, classConstructor: Identifier, privateField: Identifier): Expression;
-        createClassStaticPrivateFieldSetHelper(receiver: Expression, classConstructor: Identifier, privateField: Identifier, value: Expression): Expression;
-        createClassStaticPrivateMethodGetHelper(receiver: Expression, classConstructor: Identifier, fn: Identifier): Expression;
-        createClassStaticPrivateReadonlyHelper(receiver: Expression, value: Expression): Expression;
-        createClassStaticPrivateWriteonlyHelper(receiver: Expression): Expression;
-        createClassStaticPrivateAccessorGetHelper(receiver: Expression, classConstructor: Identifier, fn: Identifier): Expression;
-        createClassStaticPrivateAccessorSetHelper(receiver: Expression, classConstructor: Identifier, fn: Identifier, value: Expression): Expression;
+        createClassPrivateFieldGetHelper(receiver: Expression, privateField: Identifier, kind?: "a" | "m" | undefined, f?: Identifier | undefined): Expression;
```
Comment (mkubilayk) [inline src/compiler/factory/emitHelpers.ts]: I cannot suggest code changes for some reason but we can change `privateField` to `state`. Do we need to make `kind` and `f` parameters optional as they have `undefined` type explicitly?

File: src/compiler/factory/emitHelpers.ts [hunk 2]
```diff
@@ -32,21 +32,8 @@ namespace ts {
...
+        createClassPrivateFieldSetHelper(receiver: Expression, privateField: Identifier, value: Expression, kind?: "a" | "m" | undefined, f?: Identifier | undefined): Expression;
```
Comment (mkubilayk) [inline src/compiler/factory/emitHelpers.ts]: Same here.

File: src/compiler/factory/emitHelpers.ts [hunk 3]
```diff
@@ -393,75 +368,42 @@ namespace ts {
...
             context.requestEmitHelper(classPrivateFieldGetHelper);
-            return factory.createCallExpression(getUnscopedHelperName("__classPrivateFieldGet"), /*typeArguments*/ undefined, [receiver, privateField]);
+            let args;
+            if(!kind && !f) {
+                args = [receiver, state];
+            }
+            else if(kind && !f) {
+                args = [receiver, state, factory.createStringLiteral(kind)];
+            }
+            else if(!kind && f) {
```
Comment (mkubilayk) [inline src/compiler/factory/emitHelpers.ts]: `if(` -> `if (`

File: src/compiler/factory/emitHelpers.ts [hunk 4]
```diff
@@ -393,75 +368,42 @@ namespace ts {
...
+            let args;
+            if(!kind && !f) {
+                args = [receiver, state, value];
+            }
+            else if(kind && !f) {
+                args = [receiver, state, value, factory.createStringLiteral(kind)];
+            }
+            else if(!kind && f) {
+                args = [receiver, state, value, factory.createVoidZero(), f];
+            }
```
Comment (mkubilayk) [inline src/compiler/factory/emitHelpers.ts]: `if(` -> `if (`

File: src/compiler/factory/emitHelpers.ts (hunk 1 above)
Comment (mkubilayk) [reply src/compiler/factory/emitHelpers.ts]: I think Ron wanted to use `a(ccessor)`, `m(ethod)` and `f(ield)` as `kind` and eliminate `undefined` in TS 4.3+.

File: src/compiler/factory/emitHelpers.ts [hunk 5]
```diff
@@ -907,11 +849,10 @@ namespace ts {
         importName: "__classPrivateFieldGet",
         scoped: false,
         text: `
-            var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, privateMap) {
-                if (!privateMap.has(receiver)) {
-                    throw new TypeError("attempted to get private field on non-instance");
-                }
-                return privateMap.get(receiver);
+            var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
```
Comment (mkubilayk) [inline src/compiler/factory/emitHelpers.ts:889]: These helpers will be very hard to read without a reference document like Ron's gist. I guess we can ask how they want to go about it.

File: src/compiler/transformers/classFields.ts [hunk 6]
```diff
@@ -7,91 +7,38 @@ namespace ts {
...
+        state: Identifier;
+        isStatic: boolean;
     }
-    interface PrivateIdentifierStaticGetterOnly {
-        placement: PrivateIdentifierPlacement.StaticGetterOnly;
-        getterName: Identifier;
-        setterName?: undefined;
-        classConstructor: Identifier;
+    interface PrivateIdentifierInfoAccessorInfo extends PrivateIdentifierInfoBase {
+        kind: "a",
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: I think using a `const enum` for `kind` might be better for readability.

```ts
const enum Kind {
    Method = "m",
    Accessor = "a",
    Field = "f"
};
```

File: src/compiler/transformers/classFields.ts [hunk 7]
```diff
@@ -7,91 +7,38 @@ namespace ts {
...
+    interface PrivateIdentifierInfoBase {
+        state: Identifier;
+        isStatic: boolean;
     }
-    interface PrivateIdentifierStaticGetterOnly {
-        placement: PrivateIdentifierPlacement.StaticGetterOnly;
-        getterName: Identifier;
-        setterName?: undefined;
-        classConstructor: Identifier;
+    interface PrivateIdentifierInfoAccessorInfo extends PrivateIdentifierInfoBase {
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: This should be `PrivateIdentifierAccessorInfo`.

File: src/compiler/transformers/classFields.ts [hunk 8]
```diff
@@ -7,91 +7,38 @@ namespace ts {
...
-        placement: PrivateIdentifierPlacement.StaticField;
-        variableName: Identifier;
-        classConstructor: Identifier;
-    }
-    interface PrivateIdentifierStaticMethod {
-        placement: PrivateIdentifierPlacement.StaticMethod;
-        functionName: Identifier;
-        classConstructor: Identifier;
+    interface PrivateIdentifierInfoBase {
+        state: Identifier;
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: A doc string explaining the three types of identifiers for this property would be nice.

File: src/compiler/transformers/classFields.ts [hunk 9]
```diff
@@ -1074,22 +911,20 @@ namespace ts {
...
-                        case PrivateIdentifierPlacement.StaticField: {
+                        else {
                             return createPrivateStaticFieldInitializer(
-                                privateIdentifierInfo.variableName,
+                                privateIdentifierInfo.writeFunction,
                                 visitNode(property.initializer, visitor, isExpression)
                             );
                         }
-                        default:
-                            return undefined;
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts:1065]: Why did we remove this?

File: src/compiler/transformers/classFields.ts [hunk 10]
```diff
@@ -1074,22 +911,20 @@ namespace ts {
             if (shouldTransformPrivateElements && isPrivateIdentifier(propertyName)) {
                 const privateIdentifierInfo = accessPrivateIdentifier(propertyName);
                 if (privateIdentifierInfo) {
-                    switch (privateIdentifierInfo.placement) {
-                        case PrivateIdentifierPlacement.InstanceField: {
+                    if(!privateIdentifierInfo.kind) {
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: I don't think we should use `kind = undefined` to indicate fields in the transformation code.

Also `if(` -> `if (`

File: src/compiler/transformers/classFields.ts [hunk 11]
```diff
@@ -1074,22 +911,20 @@ namespace ts {
...
                                 visitNode(property.initializer, visitor, isExpression),
-                                privateIdentifierInfo.weakMapName
+                                privateIdentifierInfo.state
                             );
                         }
-                        case PrivateIdentifierPlacement.StaticField: {
+                        else {
                             return createPrivateStaticFieldInitializer(
-                                privateIdentifierInfo.variableName,
+                                privateIdentifierInfo.writeFunction,
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: I understand having a unified name helps with the creation of helper calls but I find it a bit confusing in places like this. I'm not sure if our internal data structures should be one-to-one to the emit helpers, especially considering helpers are polymorphic.

Comment (mkubilayk) [reply src/compiler/transformers/classFields.ts]: I definitely agree that there is room for simplification in the old code. But I'd prefer pushing some complexity into `createPrivateIdentifierAssignment` and `createPrivateIdentifierAccessHelper` and have `PrivateIdentifierInfo` types to have more clear properties based on `kind`.

File: src/compiler/factory/emitHelpers.ts (hunk 1 above)
Comment (dragomirtitian) [reply src/compiler/factory/emitHelpers.ts]: The helpers I saw from Ron don't actually used `field`. I think part of his suggestion was keeping the helpers backward compatible (so you could use a newer tslib with an older TS version).  I'll remove the optional, as it is always called with all arguments anyway

File: src/compiler/factory/emitHelpers.ts (hunk 3 above)
Comment (dragomirtitian) [reply src/compiler/factory/emitHelpers.ts]: Ahhhh .. that `if(` is going to be the death of me. Fixed, thanks

File: src/compiler/factory/emitHelpers.ts (hunk 5 above)
Comment (dragomirtitian) [reply src/compiler/factory/emitHelpers.ts:889]: I can copy the comments in the gist to the source code. Other helpers have extensive comments

File: src/compiler/transformers/classFields.ts (hunk 9 above)
Comment (dragomirtitian) [reply src/compiler/transformers/classFields.ts:1065]: Added back

File: src/compiler/transformers/classFields.ts (hunk 10 above)
Comment (dragomirtitian) [reply src/compiler/transformers/classFields.ts]: Changed to enum

File: src/compiler/factory/emitHelpers.ts [hunk 12]
```diff
@@ -901,184 +842,127 @@ namespace ts {
...
+     *      __classPrivateFieldSet(<any>, <constructor>, <any>, "a", <function>)
+     *
+     * Writing to a private static set accessor (when not defined, TS 4.3+):
+     *      __classPrivateFieldSet(<any>, <constructor>, <any>, "a", void 0)
+     *      NOTE: This always results in a runtime error.
+     *
+     * Writing to a private static method (TS 4.3+):
+     *      __classPrivateFieldSet(<any>, <constructor>, <any>, "m", <function>)
+     *      NOTE: This always results in a runtime error.
+     */
```
Comment (mkubilayk) [inline src/compiler/factory/emitHelpers.ts:883]: Looks great!

nit: Could remove `(existing)` as those will not make sense in the long term.

File: src/compiler/factory/emitHelpers.ts [hunk 13]
```diff
@@ -901,184 +842,127 @@ namespace ts {
...
+     *  @param state — One of the following (existing):
+     *      - A WeakMap used to store a private instance field (existing).
+     *      - A WeakSet used as an instance brand for private instance methods and accessors.
+     *      - A function value that should be the undecorated class constructor used to brand check private static fields, methods, and accessors.
+     *  @param value — The value to set (existing case).
+     *  @param kind — (optional pre TS 4.3, required for TS 4.3+) One of the following values:
+     *       - undefined — Indicates a private instance field (pre TS 4.3).
+     *       - "field" — Indicates a private field (instance or static).
+     *       - "method" — Indicates a private method (instance or static).
+     *       - "accessor" — Indicates a private accessor (instance or static).
```
Comment (mkubilayk) [inline src/compiler/factory/emitHelpers.ts]: These should probably be one-letter forms.

File: src/compiler/factory/emitHelpers.ts [hunk 14]
```diff
@@ -901,184 +842,127 @@ namespace ts {
...
+     *  @param receiver — The object from which the private member will be read (existing).
+     *  @param state — One of the following (existing):
+     *      - A WeakMap used to read a private instance field (existing).
+     *      - A WeakSet used as an instance brand for private instance methods and accessors.
+     *      - A function value that should be the undecorated class constructor used to brand check private static fields, methods, and accessors.
+     *  @param kind — (optional pre TS 4.3, required for TS 4.3+) One of the following values:
+     *      - undefined — Indicates a private instance field (pre TS 4.3).
+     *      - "field" — Indicates a private field (instance or static).
+     *      - "method" — Indicates a private method (instance or static).
+     *      - "accessor" — Indicates a private accessor (instance or static).
```
Comment (mkubilayk) [inline src/compiler/factory/emitHelpers.ts]: Same here.

File: src/compiler/transformers/classFields.ts [hunk 15]
```diff
@@ -7,95 +7,72 @@ namespace ts {
...
-    }
-    interface PrivateIdentifierStaticGetterAndSetterOnly {
-        placement: PrivateIdentifierPlacement.StaticGetterAndSetter;
-        getterName: Identifier;
-        setterName: Identifier;
-        classConstructor: Identifier;
+    interface PrivateIdentifierInfoBase {
+        /**
+         * State can contain:
+         *  - For instance field: The WeakMap that will be the storage for teh field.
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]:
```suggestion
         *  - For instance field: The WeakMap that will be the storage for the field.

```

File: src/compiler/transformers/classFields.ts [hunk 16]
```diff
@@ -7,95 +7,72 @@ namespace ts {
...
+         *  - For instance methods or accessors: The WeakSet that will be used.
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]:
```suggestion
         *  - For instance methods or accessors: The WeakSet that will be used for brand checking.

```

File: src/compiler/transformers/classFields.ts [hunk 17]
```diff
@@ -311,22 +288,14 @@ namespace ts {
...
-                    info.placement === PrivateIdentifierPlacement.StaticGetterAndSetter)) {
+            if (isGetAccessor(node) && info.kind === "a") {
                 return info.getterName;
             }
-            if (isSetAccessor(node) &&
-                (info.placement === PrivateIdentifierPlacement.InstanceSetterOnly ||
-                    info.placement === PrivateIdentifierPlacement.StaticSetterOnly ||
-                    info.placement === PrivateIdentifierPlacement.InstanceGetterAndSetter ||
-                    info.placement === PrivateIdentifierPlacement.StaticGetterAndSetter)) {
+            if (isSetAccessor(node) && info.kind === "a") {
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: This and checks above can use `PrivateIdentifierKind` enum.

File: src/compiler/transformers/classFields.ts [hunk 18]
```diff
@@ -7,95 +7,72 @@ namespace ts {
...
-        setterName: Identifier;
-        classConstructor: Identifier;
+    interface PrivateIdentifierInfoBase {
+        /**
+         * State can contain:
+         *  - For instance field: The WeakMap that will be the storage for teh field.
+         *  - For instance methods or accessors: The WeakSet that will be used.
+         *  - For static members: The constructor that will be used for brand checking.
+         */
+        state: Identifier;
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: nit: Do you think `brandCheckIdentifier` would be a better name? I know it's more than that for instance fields but `state` is a bit vague. No strong opinions though.

File: src/compiler/transformers/classFields.ts [hunk 19]
```diff
@@ -7,95 +7,72 @@ namespace ts {
...
-
-    interface PrivateIdentifierInstanceSetterOnly {
-        placement: PrivateIdentifierPlacement.InstanceSetterOnly;
-        weakSetName: Identifier;
-        setterName: Identifier;
+    interface PrivateIdentifierStaticFieldInfo extends PrivateIdentifierInfoBase {
+        kind: PrivateIdentifierKind.Field;
+        isStatic: true;
+        /**
+         * Contains the variable that will server as teh storage for the field.
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]:
```suggestion
         * Contains the variable that will server as the storage for the field.

```

File: src/compiler/transformers/classFields.ts [hunk 20]
```diff
@@ -1231,60 +1162,57 @@ namespace ts {
...
                 else if (isGetAccessorDeclaration(node)) {
                     const getterName = createHoistedVariableForPrivateName(text + "_get", node);
                     const previousInfo = env.identifiers.get(node.name.escapedText);
-                    if (previousInfo?.placement === PrivateIdentifierPlacement.StaticSetterOnly) {
-                        info = {
-                            ...previousInfo,
-                            placement: PrivateIdentifierPlacement.StaticGetterAndSetter,
-                            getterName,
-                        };
+                    if (previousInfo && previousInfo.kind === PrivateIdentifierKind.Accessor) {
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: nit: Could use optional chaining.

File: src/compiler/transformers/classFields.ts [hunk 21]
```diff
@@ -1231,60 +1162,57 @@ namespace ts {
...
                 else if (isSetAccessorDeclaration(node)) {
                     const setterName = createHoistedVariableForPrivateName(text + "_set", node);
                     const previousInfo = env.identifiers.get(node.name.escapedText);
-                    if (previousInfo?.placement === PrivateIdentifierPlacement.StaticGetterOnly) {
-                        info = {
-                            ...previousInfo,
-                            placement: PrivateIdentifierPlacement.StaticGetterAndSetter,
-                            setterName,
-                        };
+                    if (previousInfo && previousInfo.kind === PrivateIdentifierKind.Accessor) {
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: nit: Could use optional chaining.

File: src/compiler/transformers/classFields.ts [hunk 22]
```diff
@@ -1323,47 +1254,40 @@ namespace ts {
...
                     const getterName = createHoistedVariableForPrivateName(text + "_get", node);
 
-                    if (previousInfo?.placement === PrivateIdentifierPlacement.InstanceSetterOnly) {
-                        info = {
-                            placement: PrivateIdentifierPlacement.InstanceGetterAndSetter,
-                            weakSetName,
-                            getterName,
-                            setterName: previousInfo.setterName
-                        };
+                    if (previousInfo && previousInfo.kind === PrivateIdentifierKind.Accessor) {
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: nit: Could use optional chaining.

File: src/compiler/transformers/classFields.ts [hunk 23]
```diff
@@ -1323,47 +1254,40 @@ namespace ts {
...
                     const setterName = createHoistedVariableForPrivateName(text + "_set", node);
 
-                    if (previousInfo?.placement === PrivateIdentifierPlacement.InstanceGetterOnly) {
-                        info = {
-                            placement: PrivateIdentifierPlacement.InstanceGetterAndSetter,
-                            weakSetName,
-                            setterName,
-                            getterName: previousInfo.getterName
-                        };
+                    if (previousInfo && previousInfo.kind === PrivateIdentifierKind.Accessor) {
```
Comment (mkubilayk) [inline src/compiler/transformers/classFields.ts]: nit: Could use optional chaining.

**Review by mkubilayk — COMMENTED** [review]
🎉 Transform code looks good. I left some comments but they are all cosmetic changes and mostly optional. Will take a look at fixtures now.

File: src/compiler/transformers/classFields.ts (hunk 18 above)
Comment (dragomirtitian) [reply src/compiler/transformers/classFields.ts]: I actually like `brandCheckIdentifier`.. the WeakMap is a bit more for instance fields .. but it still is the brand check as well.

**Review by mkubilayk — APPROVED** [review]

**Pull request merged**